### DISCUSSION TITLE
CustomGeometrySource - Plans?

### DISCUSSION THREAD

**Discussion by BTolputt** [josxha/flutter-maplibre#130] [Ideas]
One of the things holding us back from being able to use this in a production application is the inability to use a CustomGeometrySource on Windows (&, it appears, iOS). Our apps need to run offline AND they need to pull data from a custom file format.

We have a solution working in C/C++ world using MapLibre-Native just fine but the UI is rather basic (IMGUI). Looking into Flutter, we found this project but it is missing the one (major) requirement needed for custom data vendors that cannot pre-process all of their data into PBF tiles.

Is there a technical reason this isn't implemented or is it simply a matter of not being implemented as not needed (yet)?

**Comment by BTolputt** [comment]
FWIW, if this can be done, I am more than happy to contribute effort towards this, but I don't want to climb down a rabbit hole others may have already delved into and found "Dart & CustomGeometrySource" to be incompatible.

**Comment by josxha** [comment]
> [...] I don't want to climb down a rabbit hole others may have already delved into and found "Dart & CustomGeometrySource" to be incompatible.

I can't think of any major blockers so it should be possible to implement. There just hasn't been anyone, who looked into this feature yet.
May I ask what kind of data you have that neither GeoJsonSource or VectorSource are viable options?

> [...] use a CustomGeometrySource on Windows (&, it appears, iOS)

Note, that Windows and iOS are currently not supported as platforms. iOS support is planned for the near future, Windows support is blocked by https://github.com/flutter/flutter/issues/31713. There where some experiments with embedded web views on windows in the meantime but not taken further.

**Reply by BTolputt** [reply]
>May I ask what kind of data you have that neither GeoJsonSource or VectorSource are viable options?

A GeoJsonSource is good for dynamic runtime data that you might change on the fly, but as it keeps all of that data in memory, it is not a good idea for large amounts of static data.

VectorSource sorts this out, but you need to preprocess all of the GIS data into tiles and this format can be suboptimal for other uses of that data (say road networks for route calculation). This leads to one of two solutions: you either double up on the data (one presentation layer preprocessed into tiles & the other "working" data format) or you spend a little CPU querying & clipping the data in a CustomGeometrySource. Our clients prefer the latter.

Which is neither here nor there at the moment as...
>Note, that Windows and iOS are currently not supported as platforms...

Our two main platforms for the current client are iOS & Windows, so I need to stick to a non-Flutter UI for now. Thanks though for the feedback and discussing the situation.
